src/System.Security.Cryptography.OpenSsl/src/System/Security/Cryptography/ECDsaOpenSsl.cs

```diff
@@ -278,94 +317,57 @@ private void FreeKey()
 
-                    return key;
-                }
-            }
+                if (key == null || key.IsInvalid)
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
 
-            // The KeySize property should have prevented a bad KeySize from being set.
-            Debug.Fail("GenerateKey: Unexpected KeySize: " + keySize);
-            throw new InvalidOperationException();  // This is to keep the compiler happy - we don't expect to hit this.
-        }
+                if (!Interop.Crypto.EcKeyGenerateKey(key))
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
 
-        private Lazy<SafeEcKeyHandle> _key;
+                _tempKeySize = Interop.Crypto.EcKeyGetSize(key);
+                KeySize = _tempKeySize;
+                _tempKeySize = 0;
 
-        private struct SupportedAlgorithm
-        {
-            public SupportedAlgorithm(int keySize, int nid)
-                : this()
-            {
-                KeySize = keySize;
-                Nid = nid;
+                return key;
             }
-
-            public int KeySize { get; private set; }
-            public int Nid { get; private set; }
-        }
-
-        private static readonly SupportedAlgorithm[] s_supportedAlgorithms =
-            RemoveAlgorithmsUnsupportedByOs(
-                new SupportedAlgorithm[]
-                {
-                    new SupportedAlgorithm(keySize: 224, nid: Interop.Crypto.NID_secp224r1),
-                    new SupportedAlgorithm(keySize: 256, nid: Interop.Crypto.NID_X9_62_prime256v1),
-                    new SupportedAlgorithm(keySize: 384, nid: Interop.Crypto.NID_secp384r1),
-                    new SupportedAlgorithm(keySize: 521, nid: Interop.Crypto.NID_secp521r1),
-                }
-            );
-
-        private static SupportedAlgorithm[] RemoveAlgorithmsUnsupportedByOs(SupportedAlgorithm[] supportedAlgorithms)
-        {
-            List<SupportedAlgorithm> filteredSupportedAlgorithms = new List<SupportedAlgorithm>(supportedAlgorithms.Length);
-            foreach (SupportedAlgorithm supportedAlgorithm in supportedAlgorithms)
+            else
             {
-                int nid = supportedAlgorithm.Nid;
-                using (SafeEcKeyHandle key = Interop.Crypto.EcKeyCreateByCurveName(nid))
+                int nid = 0;
+                int keySize = KeySize;
+                switch (keySize)
                 {
-                    if (key != null && !key.IsInvalid)
-                    {
-                        filteredSupportedAlgorithms.Add(supportedAlgorithm);
-                    }
+                    case 224: nid = Interop.Crypto.NID_secp224r1;break;
+                    case 256: nid = Interop.Crypto.NID_X9_62_prime256v1; break;
+                    case 384: nid = Interop.Crypto.NID_secp384r1; break;
+                    case 521: nid = Interop.Crypto.NID_secp521r1; break;
```

Comment:
Rather than this sort of fallback, we could make setting KeySize to one of these three (dropping 224) values set the curve to the appropriate named curve by OID; and drop the manual NID mapping.